### PR TITLE
Spark 3.5: Remove constructor from parameterized base class

### PR DESCRIPTION
**Pull request by nastra**
After reviewing/merging https://github.com/apache/iceberg/pull/9342 I realized that parameterized base classes should not have a constructor but rather have their parameters set via `@Parameter`. All other setup code should go into a `@BeforeEach` method to properly initialize the test environment

### CODE REVIEW

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/CatalogTestBase.java [hunk 1]
```diff
@@ -30,7 +29,7 @@ public abstract class CatalogTestBase extends TestBaseWithCatalog {
 
   // these parameters are broken out to avoid changes that need to modify lots of test suites
   @Parameters(name = "catalogName = {0}, implementation = {1}, config = {2}")
-  public static Object[][] parameters() {
+  protected static Object[][] parameters() {
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/CatalogTestBase.java:32]: JUnit4 required this to be public but for JUnit5 we can make it protected

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/TestBaseWithCatalog.java [hunk 2]
```diff
@@ -35,11 +38,25 @@
...
 
+@ExtendWith(ParameterizedTestExtension.class)
 public abstract class TestBaseWithCatalog extends TestBase {
   protected static File warehouse = null;
 
+  @Parameters(name = "catalogName = {0}, implementation = {1}, config = {2}")
+  protected static Object[][] parameters() {
+    return new Object[][] {
+      {
+        SparkCatalogConfig.HADOOP.catalogName(),
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/TestBaseWithCatalog.java:53]: all of the subclasses of this test that didn't specify any particular parameter were running with this configuration (this is also the configuration that was initialized in the default constructors in L70

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestDataFrameWriterV2.java [hunk 3]
```diff
@@ -50,7 +50,7 @@ public void removeTables() {
     sql("DROP TABLE IF EXISTS %s", tableName);
   }
 
-  @Test
+  @TestTemplate
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestDataFrameWriterV2.java:53]: all of the subclasses of `TestBaseWithCatalog` are parameterized tests

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkStagedScan.java [hunk 4]
```diff
@@ -39,11 +38,6 @@
 
 public class TestSparkStagedScan extends CatalogTestBase {
 
-  public TestSparkStagedScan(
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkStagedScan.java:42]: this and the class below were running with the three different catalog configurations defined in `CatalogTestBase` and I've verified that this is still the case

**Review by Fokko — APPROVED** [review]
Great improvement @nastra! Looking good

**Pull request merged**